### PR TITLE
Add `RuleConfig`

### PR DESCRIPTION
**Pull request by mcabbott**
This wants to add methods for `rrule_via_ad` and `frule_via_ad` from ChainRules. 

~~Need to figure out how to test this; so far `rrule_via_ad` does seem to work.~~ Not sure anything is in the right place.

### CODE REVIEW

**Comment by simeonschaub** [comment]
Testing should now be set up, so you'll probably want to rebase this.

**Comment by simeonschaub** [comment]
Would it be possible to add a test specifically making use of this?

**Comment by mcabbott** [comment]
Yes was just trying to figure out how (and where). I've added something, but my `frule_via_ad` definition is clearly completely wrong.

File: src/runtime.jl [hunk 1]
```diff
@@ -1,5 +1,7 @@
 using ChainRulesCore
 
+struct DiffractorRuleConfig <: RuleConfig{Union{HasReverseMode,HasForwardsMode}} end
```
Comment (mcabbott) [inline src/runtime.jl:3]: This doesn't seem a great fit for the terse style of this package, better ideas are welcome.

Comment (simeonschaub) [reply src/runtime.jl:3]: You mean the name? Seems fine to me.

Comment (mcabbott) [reply src/runtime.jl:3]: Yes, sorry somehow I lost the word.

File: test/runtests.jl [hunk 2]
```diff
@@ -1,8 +1,8 @@
 using Diffractor
-using Diffractor: var"'", ∂⃖
+using Diffractor: var"'", ∂⃖, DiffractorRuleConfig
 using ChainRules
 using ChainRulesCore
-using ChainRules: ZeroTangent, NoTangent
+using ChainRules: ZeroTangent, NoTangent, frule_via_ad, rrule_via_ad
```
Comment (oxinabox) [inline test/runtests.jl:5]: These should be loaded from ChainRulesCore

**Review by oxinabox — APPROVED** [review]
You can test this using ChainRulesTestUtils if you wanted.
It supports testing these cases.

This basically LGTM.
I see no reason not to merge it once tests pass

File: src/stage1/forward.jl [hunk 3]
```diff
@@ -118,6 +118,13 @@ function (::∂☆internal{1})(args::AbstractTangentBundle{1}...)
     end
 end
 
+function ChainRulesCore.frule_via_ad(::DiffractorRuleConfig, partials, args...)
+    tangents = map(partials, args) do p, a
+        TangentBundle{1}(a, (p,))
+    end
+    ∂☆internal{1}()(tangents...)
+end
```
Comment (mcabbott) [inline src/stage1/forward.jl:126]: I don't see what's going wrong here. The error is this:
```
julia> z45, delta45 = frule_via_ad(DiffractorRuleConfig(), (ZeroTangent(),1), sqrt, 2.0)
ERROR: MethodError: no method matching iterate(::Diffractor.TangentBundle{1, Float64, Tuple{Float64}})
Closest candidates are:
  iterate(::Union{LinRange, StepRangeLen}) at range.jl:737
  iterate(::Union{LinRange, StepRangeLen}, ::Int64) at range.jl:737
  iterate(::T) where T<:Union{Base.KeySet{<:Any, <:Dict}, Base.ValueIterator{<:Dict}} at dict.jl:695
  ...
Stacktrace:
 [1] indexed_iterate(I::Diffractor.TangentBundle{1, Float64, Tuple{Float64}}, i::Int64)
   @ Base ./tuple.jl:92
 [2] top-level scope
   @ REPL[54]:1
```
It certainly calls `shuffle_base(r) = TangentBundle{1}(r[1], (r[2],))` (10 lines up) but after that I don't know.

Comment (simeonschaub) [reply src/stage1/forward.jl:126]: I don't think you can destructure a Tangent Bundle like this.

Comment (mcabbott) [reply src/stage1/forward.jl:126]: But what is destructuring it? The thing I construct called `tangents` is just a Tuple, and looks acceptable to `∂☆internal{1}()`. From the type `tangents[2]` is being iterated but I don't see where.

Comment (simeonschaub) [reply src/stage1/forward.jl:126]: My suspicion is that it's you doing `z45, delta45 = frule_via_ad(...)` inside the test, but I could be wrong.

Comment (mcabbott) [reply src/stage1/forward.jl:126]: Oh right, ok. That would explain the short stack trace! This must return a tuple but surely I can figure out how to convert it.

**Comment by simeonschaub** [comment]
CI failure is https://github.com/JuliaLang/julia/pull/42125, will push a fix shortly if no one beats me to it.

**Comment by simeonschaub** [comment]
Just as a small comment: I'd generally always squash-merge PRs like this unless the commits are actually standalone changes.

**Comment by mcabbott** [comment]
Sorry, will do, next time.

**Pull request merged**